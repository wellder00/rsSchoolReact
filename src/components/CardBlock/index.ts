export { default as CardBlock } from './CardBlock';
