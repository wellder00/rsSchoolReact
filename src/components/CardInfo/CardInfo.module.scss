@import '../../styles/variables.scss';

.characterInfo {
  position: relative;
  margin-top: 20px;
  flex: 1;
}

.infoWrap {
  position: sticky;
  top: 20px;
  background: $card-color;
  text-align: center;
  font-size: 18px;
  border: 1px solid black;
  border-radius: 15px;
}

.title {
  margin: 20px 0 0 0;
  filter: drop-shadow(2px 1px 2px red);
  color: orange;
}

.blockInfo {
  padding: 10px 0;
}

.blockInfo div {
  padding: 2px 0;
}

.image {
  width: 300px;
  border-radius: 15px;
  @media (max-width: 1000px) {
    width: 200px;
  }
}

.listWrap {
  display: flex;
  justify-content: center;
}

.list {
  list-style: none;
  text-align: start;
  padding: 0 4px;
}
