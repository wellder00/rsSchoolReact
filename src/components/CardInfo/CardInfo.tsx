import { Suspense } from 'react';
import { useLoaderData, Link, LoaderFunction, defer, Await } from 'react-router-dom';

import styles from './CardInfo.module.scss';

import { Character } from '../../types/interfaces';
import { pokemonAPI } from '../../utils/constants/api';
import { Button } from '@components/Button';
import { Loader } from '@components/Loader';

async function getCharacter(id: number) {
  if (!id || typeof id !== 'number') {
    throw new Error('Invalid character ID');
  }
  const response = await fetch(`${pokemonAPI}/${id}`);
  if (!response.ok) {
    throw new Error('ERROR HTTP: ' + response.status);
  }
  return response.json();
}

export const loader: LoaderFunction = async ({ params }) => {
  if (params && params.characterId !== undefined) {
    return defer({
      character: getCharacter(+params.characterId),
    });
  }
};

interface CharacterData {
  character: Character;
}

const CardInfo = () => {
  const { character } = useLoaderData() as CharacterData;

  return (
    <div className={styles.characterInfo}>
      <div className={styles.infoWrap}>
        <Suspense fallback={<Loader />}>
          <Await resolve={character}>
            {(person) => (
              <>
                <h3 className={styles.title}>{person.name}</h3>
                <div className={styles.blockInfo}>
                  <div>Status: {person.status.toLowerCase()}</div>
                  <div>Species: {person.species.toLowerCase()}</div>
                  <div>Gender: {person.gender.toLowerCase()}</div>
                  <div>Location: {person.location?.name}</div>
                  <div>Origin: {person.origin?.name}</div>
                </div>
                <img className={styles.image} src={person.image} alt="image person" />
                <div>
                  <Link to={'/'}>
                    <Button className={'backButton'}>Back</Button>
                  </Link>
                </div>
              </>
            )}
          </Await>
        </Suspense>
      </div>
    </div>
  );
};

export default CardInfo;
