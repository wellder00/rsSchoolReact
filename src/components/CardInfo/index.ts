export { default as CardInfo } from './CardInfo';
