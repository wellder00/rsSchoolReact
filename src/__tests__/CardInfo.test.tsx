import { render, waitFor } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import { CardInfo } from '../components/CardInfo';
import { ditto } from './mocks/PokemonData';

test('CardInfo', async () => {
  const routes = [
    {
      path: '/about_character/8',
      element: <CardInfo />,
      loader: () => ditto,
    },
  ];

  const router = createMemoryRouter(routes, { initialEntries: ['/about_character/8'] });

  const { asFragment } = render(<RouterProvider router={router} />);

  await waitFor(() => {
    console.log(ditto);
    expect(asFragment()).toMatchSnapshot();
  });
});
