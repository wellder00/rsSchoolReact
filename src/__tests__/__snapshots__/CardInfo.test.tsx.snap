// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardInfo Component Find card weight 1`] = `
<DocumentFragment>
  <div
    class="characterInfo"
  >
    <div
      class="infoWrap"
    >
      <h3
        class="title"
      >
        ditto
      </h3>
      <div
        class="blockInfo"
      >
        <div>
          Weight: 40
        </div>
        <div>
          Species: ditto
        </div>
        <div
          class="listWrap"
        >
          <ul
            class="list"
          >
            <li>
              defense: string
            </li>
          </ul>
        </div>
      </div>
      <img
        alt="image person"
        class="image"
        data-nimg="1"
        decoding="async"
        height="300"
        loading="lazy"
        src="/_next/image?url=%2Fditto&w=640&q=75"
        srcset="/_next/image?url=%2Fditto&w=384&q=75 1x, /_next/image?url=%2Fditto&w=640&q=75 2x"
        style="color: transparent;"
        width="300"
      />
      <div>
        <a
          href="/?page=1&limit=3&offset=0"
        >
          <button
            class="backButton button"
          >
            Back
          </button>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
